[SYCL] Adds compile-time properties

steffenlarsen opened the pull request:
Implements the [compile-time property list extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/PropertyList/SYCL_EXT_ONEAPI_property_list.asciidoc) adding the `sycl::ext::oneapi::properties` and `sycl::ext::oneapi::property_value` classes. These allow for compile-time inspection of properties passed to SYCL objects. These changes do not add support for the new property lists to any existing SYCL classes however.

steffenlarsen commented:
@gmlueck | @rolandschulz | @Pennycook - It seems that MSVC is complaining because `get_property` has both static and non-static overloads. It appears to be a problem we have had with `accessor_property_list` before as well: https://www.intel.com/content/www/us/en/developer/articles/troubleshooting/c2686-when-using-msvc-as-host-compiler.html

dm-vodopyanov commented:
@steffenlarsen, can you try exactly the same syntax of `std::enable_if_t` from https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/accessor_property_list.hpp? It turned out this combination of `std::enable_t` syntax works for MSVC in accessor_property_list. I hope it helps.

cperkinsintel commented on sycl/include/sycl/ext/oneapi/property_list/property_value.hpp:
```diff
@@ -0,0 +1,105 @@
+//     property `foo`, there should be a definition
+//     `inline constexpr foo::value_t foo_v`.
+//  5. Specialize `sycl::is_property` and `sycl::is_property_of` for the
+//     property class.
```
Since there are a lot of steps, maybe give some code example in the comments, or refer the reader to one that is properly implemented elsewhere.

steffenlarsen replied:
I have moved this explanation to the new file `sycl/include/sycl/ext/oneapi/property_list/properties.hpp`, added a how-to for adding runtime properties, and examples to both. What do you think?

steffenlarsen commented:
> @steffenlarsen, can you try exactly the same syntax of `std::enable_if_t` from https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/accessor_property_list.hpp? It turned out this combination of `std::enable_t` syntax works for MSVC in accessor_property_list. This patch made it work: #4240. I hope it helps for property_list.

Thank you! That seems to have done the trick.

steffenlarsen commented:
Recent changes:

- Moved the implementation into `sycl::ext::oneapi::experimental`.
- Adjusted for updates in https://github.com/intel/llvm/pull/5338.

steffenlarsen commented:
https://github.com/intel/llvm/pull/5338 has been merged and https://github.com/intel/llvm/pull/5410 moves the API into `experimental` namespace.

This PR is ready for reviews.

steffenlarsen commented:
> This PR is ready for reviews.

@cperkinsintel @rolandschulz @Pennycook @gmlueck @dm-vodopyanov @romanovvlad @AaronBallman - Friendly ping for awareness.

romanovvlad commented on sycl/include/CL/sycl/types.hpp:
```diff
@@ -2399,6 +2399,12 @@ struct is_device_copyable<
 
+// Specializations of device copyable should propagate onto constants.
+template <typename T>
+struct is_device_copyable<
```
@gmlueck
Could you please comment on assuming that if `T` is device copyable then `const T` is device copyable as well? What about `volatile`?

gmlueck replied:
Strictly speaking, the SYCL spec does not say that the implementation automatically assumes that "const T" is device copyable if "T" is device copyable.  According to the current wording of the spec, the application is responsible for specializing `is_device_copyable` for all cv-qualified versions of `T` that it wants to declare as "device copyable".

Therefore, it's probably better not to add these lines.  If the goal is to declare that `sycl::ext::oneapi::experimental::properties` is device copyable, it would be better to add specializations just for that type (and any cv-qualified versions we care about).

steffenlarsen replied:
Makes sense to me. I have removed this and specialized `is_device_copyable` for `const properties` instead. We may want to do this for `std::tuple` and `sycl::marray` as well, but I don't think this PR is the right place for that.

keryell replied:
I am still puzzled by some use-cases where the device copy-ability would depend on CV qualification.
But yeah... keep the door open. :-)

romanovvlad commented on sycl/include/CL/sycl/types.hpp:
```diff
@@ -2399,6 +2399,12 @@ struct is_device_copyable<
+// Specializations of device copyable should propagate onto constants.
+template <typename T>
+struct is_device_copyable<
+    const T, std::enable_if_t<!std::is_trivially_copyable<const T>::value>>
```
Could you please add a comment explaining why this `enable_if` is needed?

steffenlarsen replied:
This has been removed.

romanovvlad commented on sycl/include/sycl/ext/oneapi/properties/property_utils.hpp:
```diff
@@ -0,0 +1,221 @@
+template <typename T> struct AllUnique<std::tuple<T>> : std::true_type {};
+template <typename L, typename R, typename... Rest>
+struct AllUnique<std::tuple<L, R, Rest...>>
+    : sycl::detail::conditional_t<PropertyID<L>::value != PropertyID<R>::value,
```
It seems it checks that neighbors have unique ids rather than that all types have unique ids.

steffenlarsen replied:
You are right. This assumes that the properties have been sorted first, in which case neighbor-comparison is enough to ensure all IDs are unique. I have moved this to be among the other sorting-related traits and renamed it to `SortedAllUnique`.

romanovvlad reviewed (commented):
LGTM

psamolysov-intel commented on sycl/test/extensions/properties/properties_is_property_value.cpp:
```diff
@@ -0,0 +1,90 @@
+                NotAPropertyKey, sycl::queue>::value);
+
+  return 0;
+}
```
No new line at the end of file.

steffenlarsen replied:
```suggestion
}

```

psamolysov-intel commented on sycl/test/extensions/properties/properties_is_property_key.cpp:
```diff
@@ -0,0 +1,91 @@
+                                                           sycl::queue>::value);
+
+  return 0;
+}
```
No new line at the end of file.

steffenlarsen replied:
```suggestion
}

```

psamolysov-intel reviewed (commented):
Just a couple of warnings from GitHub

romanovvlad commented:
@gmlueck @rolandschulz @cperkinsintel @psamolysov-intel 
Could you please tell if the patch looks OK to you and if you are planning/not planning to review it?

psamolysov-intel commented:
I'm not going to dive in details but from high-level perspective the patch looks good for me.

gmlueck commented:
> Could you please tell if the patch looks OK to you and if you are planning/not planning to review it?

Of the LEWG members, @rolandschulz was most involved in the implementation details of compile-time properties.  I was not planning to review in detail, but @rolandschulz might want to.

bader commented:
@rolandschulz, could you take a look, please?

bader commented:
@rolandschulz, ping.

steffenlarsen commented:
@rolandschulz - Could you please clarify if you plan on reviewing this? If not, @Pennycook would you mind having a look?

Pennycook commented:
> @rolandschulz - Could you please clarify if you plan on reviewing this? If not, @Pennycook would you mind having a look?

Like Greg, I wasn't planning to do an in-depth review here.

I did take a look when the PR was opened, but the template wizardry being employed is sufficiently advanced that I wouldn't feel comfortable reviewing for things like impact on compile-time, etc.

FWIW, what I understood looks good, and I appreciated the "How to add a new property" comments -- I think those will be very useful. 👍

cperkinsintel reviewed (approved):
Lime Gin Tastes Metallic!

cperkinsintel commented on sycl/include/sycl/ext/oneapi/properties/property.hpp:
```diff
@@ -0,0 +1,174 @@
+//     `sycl::ext::oneapi::experimental::detail::PropKind` representing the new
+//     property. Increment
+//     `sycl::ext::oneapi::experimental::detail::PropKind::PropKindSize`
+//  2. Define property key class with `value_t` that must be `property_value`
```
can these new property keys be in other namespaces?

steffenlarsen replied:
Sure. There is no requirement on the namespace of the properties. It would probably be preferable to have them in a namespace under `sycl::ext::oneapi::experimental` while it is still experimental, but since it is user-facing it should not be a requirement from the implementation.

cperkinsintel commented on sycl/include/sycl/ext/oneapi/properties/property.hpp:
```diff
@@ -0,0 +1,174 @@
+//     class, e.g. for a property `foo`, there should be a definition
+//     `inline constexpr foo_key::value_t foo`.
+//  4. Specialize `sycl::ext::oneapi::experimental::is_property` and
+//     `sycl::ext::oneapi::experimental::is_property_of` for the property key
```
do we specialize `is_property` for the key? Or do we specialize `is_property_key` ?

Same applies to `is_property_of` vs `is_property_key_of` ?

steffenlarsen replied:
Good eye. This was missed with the renaming and should be `is_property_key` and `is_property_key_of`.

cperkinsintel commented on sycl/test/extensions/properties/mock_compile_time_properties.hpp:
```diff
@@ -0,0 +1,140 @@
+namespace detail {
+template <> struct PropertyToKind<bar_key> {
+  static constexpr PropKind Kind =
+      static_cast<enum PropKind>(PropKind::PropKindSize + 0);
```
I'm guessing the awkward `PropKindSize + X` constructs here is because we aren't actually making modifications to `property.hpp`, correct?

steffenlarsen replied:
Exactly. These properties are not "real" so we make their `PropKind` outside the usual range to avoid confusing them with actual properties.

cperkinsintel commented on sycl/include/sycl/ext/oneapi/properties/property.hpp:
```diff
@@ -0,0 +1,174 @@
+// (4.)
+template <> struct is_property<bar_key> : std::true_type {};
+// Replace SYCL_OBJ with the SYCL object to support the property.
+template <> struct is_property_of<bar_key, SYCL_OBJ> : std::true_type {};
```
what if the SYCL_OBJ class itself requires template specialization?  Use a template template? Proxy? Will we have order-of-evaluation issues?

steffenlarsen replied:
`is_property_of` is already a thing in SYCL 2020. For example, if we specialize for buffer we have something long the lines of
```c++
template <typename T, int Dimensions, typename AllocatorT>
struct is_property_key_of<bar_key, buffer<T, Dimensions, AllocatorT, void>> : std::true_type {};
```
I don't believe there should be any issues with that.

steffenlarsen commented:
Failures were unrelated. They have been reported and testing has been restarted.

steffenlarsen commented:
@romanovvlad & @psamolysov-intel & @cperkinsintel - There has been some minor corrections to the how-to. Would you mind having another look?

psamolysov-intel reviewed (approved):
I didn't review deeply, but what I see in the latest 3 commits looks good for me. Thanks, @steffenlarsen!

steffenlarsen commented:
@intel/llvm-gatekeepers - Remaining questions have been answered. This should be ready for merge.

bader commented:
@steffenlarsen, [post-commit](https://github.com/intel/llvm/runs/5358967864?check_suite_focus=true) failed with
```
/home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/property_value.hpp:54:51: error: unused parameter 'LHS' [-Werror,-Wunused-parameter]
operator==(const property_value<PropertyT, A...> &LHS,
                                                  ^
/home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/property_value.hpp:55:51: error: unused parameter 'RHS' [-Werror,-Wunused-parameter]
           const property_value<PropertyT, B...> &RHS) {
                                                  ^
/home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/property_value.hpp:62:51: error: unused parameter 'LHS' [-Werror,-Wunused-parameter]
operator!=(const property_value<PropertyT, A...> &LHS,
                                                  ^
/home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/property_value.hpp:63:51: error: unused parameter 'RHS' [-Werror,-Wunused-parameter]
           const property_value<PropertyT, B...> &RHS) {
                                                  ^
In file included from /home/runner/work/llvm/llvm/src/sycl/source/backend/level_zero.cpp:9:
In file included from /home/runner/work/llvm/llvm/src/sycl/include/CL/sycl.hpp:67:
/home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/properties.hpp:99:[42](https://github.com/intel/llvm/runs/5358967864?check_suite_focus=true#step:9:42): error: unused parameter 'PropertyValues' [-Werror,-Wunused-parameter]
  Extract(std::tuple<PropertyValueTs...> PropertyValues) {
                                         ^
5 errors generated.
```
Could you fix these issues, please?

steffenlarsen commented:
> @steffenlarsen, [post-commit](https://github.com/intel/llvm/runs/5358967864?check_suite_focus=true) failed with
> 
> ```
> /home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/property_value.hpp:54:51: error: unused parameter 'LHS' [-Werror,-Wunused-parameter]
> operator==(const property_value<PropertyT, A...> &LHS,
>                                                   ^
> /home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/property_value.hpp:55:51: error: unused parameter 'RHS' [-Werror,-Wunused-parameter]
>            const property_value<PropertyT, B...> &RHS) {
>                                                   ^
> /home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/property_value.hpp:62:51: error: unused parameter 'LHS' [-Werror,-Wunused-parameter]
> operator!=(const property_value<PropertyT, A...> &LHS,
>                                                   ^
> /home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/property_value.hpp:63:51: error: unused parameter 'RHS' [-Werror,-Wunused-parameter]
>            const property_value<PropertyT, B...> &RHS) {
>                                                   ^
> In file included from /home/runner/work/llvm/llvm/src/sycl/source/backend/level_zero.cpp:9:
> In file included from /home/runner/work/llvm/llvm/src/sycl/include/CL/sycl.hpp:67:
> /home/runner/work/llvm/llvm/build/include/sycl/ext/oneapi/properties/properties.hpp:99:[42](https://github.com/intel/llvm/runs/5358967864?check_suite_focus=true#step:9:42): error: unused parameter 'PropertyValues' [-Werror,-Wunused-parameter]
>   Extract(std::tuple<PropertyValueTs...> PropertyValues) {
>                                          ^
> 5 errors generated.
> ```
> 
> Could you fix these issues, please?

https://github.com/intel/llvm/pull/5684 should fix it.
